SSA/Projects/Tensor1D/Tensor1D.lean

```diff
@@ -294,142 +294,74 @@ theorem Tensor1d.tile [Inhabited α] (t : Tensor1d α) (SIZE : 4 ∣ t.size) (f
     sorry
 }
 
-
-
-
-
-
-
 namespace ArithScfLinalg
 
-open Val
 
-inductive Op
+
+inductive ExOp
 | add
 | const (v: Nat)
 | sub
--- | mul
--- | run
--- | for_
--- | if_
--- | fold1d -- fold
 | map1d
 | extract1d
--- | fill
--- | transpose
-
-inductive BaseType
-| int : BaseType
-| nat : BaseType
-| tensor1d  : BaseType
-| tensor2d : BaseType
+deriving DecidableEq
+
+inductive ExTy
+| int : ExTy
+| nat : ExTy
+| tensor1d  : ExTy
 deriving DecidableEq, Inhabited
 
-instance : Goedel BaseType where
+instance : Goedel ExTy where
   toType
   | .int => Int
   | .nat => Nat
   | .tensor1d => Tensor1d Int
-  | .tensor2d => Tensor2d Int
-
-
-abbrev UserType := SSA.UserType BaseType
-
--- Can we get rid of the code repetition here? (not that copilot has any trouble completing this)
-@[simp]
-def argUserType : Op → UserType
-| Op.add => .pair (.base BaseType.int) (.base BaseType.int)
-| Op.sub => .pair (.base BaseType.int) (.base BaseType.int)
-| Op.map1d => .base BaseType.tensor1d
-| Op.extract1d => .triple (.base BaseType.tensor1d) (.base BaseType.nat) (.base BaseType.nat)
-| Op.const _ => .unit
-
-@[simp]
-def outUserType : Op → UserType
-| Op.add => .base (BaseType.int)
-| Op.sub => .base (BaseType.int)
-| Op.const _ => .base (BaseType.nat)
-| Op.map1d =>  .base BaseType.tensor1d
-| Op.extract1d =>  .base BaseType.tensor1d
-
-@[simp]
-def rgnDom : Op → UserType
-| Op.add => .unit
-| Op.sub => .unit
-| Op.const _ => .unit
-| Op.map1d => .base BaseType.int
-| Op.extract1d => .unit
-
-@[simp]
-def rgnCod : Op → UserType
-| Op.add => .unit
-| Op.sub => .unit
-| Op.const _ => .unit
-| Op.map1d => .base BaseType.int
-| Op.extract1d => .unit
-
-def eval (o : Op)
-  (arg: Goedel.toType (argUserType o))
-  (_rgn : (Goedel.toType (rgnDom o) → (Goedel.toType (rgnCod o)))) :
-  (Goedel.toType (outUserType o)) :=
-  match o with
-  | .const v => v
-  | .add =>
-    let (x, y) := arg;
-    let x : Int := x;
-    let y : Int := y;
-    x + y
-  | .sub =>
-    let (x, y) := arg;
+
+instance : OpSignature ExOp ExTy where
+  outTy : ExOp → ExTy
+  | .add => .int
+  | .sub => .int
+  | .const _ => .nat
+  | .map1d =>  .tensor1d
+  | .extract1d =>  .tensor1d
+  sig  : ExOp → List ExTy
+  | .add => [.int, .int]
+  | .sub => [.int, .int]
+  | .map1d => [.tensor1d]
+  | .extract1d => [.tensor1d, .nat, .nat]
+  | .const _ => []
+  
+  regSig
+  | .map1d => [([.int], .int)]
+  | _ => []
+
+/-
+-- Error: unknown free variable: _kernel_fresh.459
+@[reducible]
+instance : OpDenote ExOp ExTy where
+  denote
+  | .const v, _, _ => v
+  | .add, (.cons x (.cons y nil)), _ =>
+      let x : Int := x;
+      let y : Int := y
+      x + y
+  | .sub, (.cons x (.cons y nil)), _ =>
     let x : Int := x;
```

Comment:
does this add implicit coercions, or force `x` and `y` to be `.int`s? why don't you use the `toTuple` function? I guess it's a style question but I'd find it more readable

Reply:
It forces `x` and `y` to be `.int`. I played around with other ways of doing it, including `toTuple`, but I was unable to convince `Lean` that `Int` was defeq to `[[ExTy.int]]` without this kind of explicit let-shadowing.

Reply:
feel invited to give it a go!